### ISSUE TITLE
Dompdf right border gone 

### ISSUE THREAD

**Issue by A-Lim** [dompdf/dompdf#1033]
The border of my table will become hidden whenever i use the styling bold and background color light gray. This there any work around for this?
this is without styling 

![without styling](https://cloud.githubusercontent.com/assets/14160760/10194399/0a31a61e-67bb-11e5-9e19-b36f40002a20.PNG)

when i change the background color to lightgray

![background](https://cloud.githubusercontent.com/assets/14160760/10194408/1e37875a-67bb-11e5-8992-2fcd013de933.PNG)

when I bold and change background color to lightgray

![background and bold](https://cloud.githubusercontent.com/assets/14160760/10194436/3c1a68aa-67bb-11e5-86c9-94b7eed6a46e.PNG)

**Comment by bsweeney** [comment]
Can you post some sample HTML so we can test?

**Comment by AlexiusLim** [comment]
Here you go. Cant upload a HTML file so I'll just upload the txt file.

**Comment by AlexiusLim** [comment]
Here you go.

```
<html>
<head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
</head>
<body>
<div>
    <table border="1" cellspacing="0" cellpadding="0" style="width:100%">
        <tbody>
        <tr>
            <td rowspan="2">
                <p>Test</p>
            </td>
            <td colspan="3">
                <p>Test</p>
            </td>
            <td rowspan="2">
                <p>Test</p>
            </td>
        </tr>
        <tr>
            <td>
                <p>Test</p>
            </td>
            <td>
                <p>Test</p>
            </td>
            <td style="background-color: lightgray">
                <p><strong>border right<strong></strong></p>
            </td>
        </tr>
        </tbody>
    </table>
</div>

</body>
</html>
```

**Comment by bsweeney** [comment]
That sample doesn't quite reproduce the problem for me, but it's close enough. There appears to be some bit of overlap going on between the cell and the adjacent border. Since you have no border spacing can you try styling the table with `border-collapse: collapse;` to see if it works around the issue?

**Comment by AlexiusLim** [comment]
Tried that and it became as what u said. Some overlapping.

**Comment by bsweeney** [comment]
It looks like the table cell in the second row (Cell A) next to the one with rowspan from the first row (Cell B) isn't correctly accounting for that cell. Because Cell A is placed in the document after Cell B it  will be rendered above Cell B in the stack. And since Cell A isn't correctly accounting for Cell B it's overlapping the boundary between the two cells.

I can't find a good way to work around the issue at this time. In your particular document, since you have no padding on the cells, if you apply the background color on a div around the content rather than on the cell it should appear as expected.

```
<td>
  <div style="background-color: lightgray">
    <p><strong>border right</strong></p>
  </div>
</td>
```

**Comment by DooRooSA** [comment]
I am getting the same problem. PDF attached, if you look at the blue row, between the two cells named "FSTD
[document.pdf (35).pdf](https://github.com/dompdf/dompdf/files/149659/document.pdf.35.pdf)
"

**Comment by DooRooSA** [comment]
Any progress on fixing this yet? My project is ready for launch and this affects me quite a bit

**Comment by bsweeney** [comment]
nothing yet

**Comment by DooRooSA** [comment]
After some testing, I've found it happens as soon as I define a background color to the TH cell

**Comment by DooRooSA** [comment]
I've also just noticed, that on my PDF's it only happens when the cell to the right has a rowspan > 1

**Comment by bsweeney** [comment]
@DooRooSA correct. Take a look at my [previous comment](#issuecomment-145194785) where I offered a potential work-around until we can address the problem.

**Comment by bsweeney** [comment]
Some additional info I wrote for https://stackoverflow.com/a/56640209/264628

---

This is a rendering issue with Dompdf versions 0.8.3 and earlier. The core of the issue is the rendering order of the table cells in combination with the technique used to render table cell borders and backgrounds when the borders are collapsed.

First, as with everything in Dompdf earlier elements are rendered before later elements in the document structure. When a cell spans rows it is rendered prior to cells in the following rows. This is important because of the stacking order of objects in the PDF (later objects are rendered above earlier ones).

Second, cell borders are rendered by a single cell within a grouping of any two adjacent cells. For horizontally adjacent cells the cell on the right renders the left border. For vertically adjacent cells the cell on the bottom renders the top border.

Third, cell backgrounds are rendered to the edge of the cell without consideration of borders.

Now take these three facts into account in consideration of a row-spanned cell that is to the right of other table cells. The left border of the row-spanned cell will be rendered. The adjacent cells in the following row will not render a border, and any background is rendered the full width of the cell. Because of this the later cell's background will render over top of the border of the row-spanned cell.

You can see this issue in the following example (when run in, for example, Dompdf 0.8.3):

    <head>
        <title>Row-span overlap</title>
        <style>
            table {
                border-collapse: collapse;
            }
            table td {
                border: 1px solid red;
                background-color: #0000ff66;
            }
            .rowspan {
                border-left-width: 10px;
            }
        </style>
    </head>
    
    <body>
        <table>
            <tbody>
                <tr>
                    <td>1</td>
                    <td rowspan="2">2</td>
                </tr>
                <tr>
                    <td>3</td>
                </tr>
            </tbody>
        </table>
    </html>

The issue will be addressed in Dompdf 0.8.4. The primary change to address the bug was to render the background within the border.

**Comment by vainiusr** [comment]
Even though in stackoverflow it's declared to be addressed in version `0.8.4`, it's still an issue in version `0.8.5`

![image](https://user-images.githubusercontent.com/54093800/88149758-2c67e600-cc09-11ea-9c8e-3fe6251368a3.png)

**Comment by bsweeney** [comment]
@vainiusr the example we were working against works OK. If you have found another use case we'll need to see sample HTML + CSS so we can confirm and work on a fix.
